README.rst
===========
metlog-hive
===========

metlog-hive is a set of extensions for Apache Hive to provide richer query support
for JSON structures that Metlog produces. 

Currently we provide a User Defined Table Function (UDTF)
`exjson_tuple` that supports indexing into a JSON structure at an
arbitrary depth with support for both dictionaries and list
structures.  This is meant to be a drop in replacement for calls to
the standard Hive `json_tuple` UDTF.

More information about how Mozilla Services is using Metlog (including what is
being used for a router and what endpoints are in use / planning to be used)
can be found on the relevant `spec page
<https://wiki.mozilla.org/Services/Sagrada/Metlog>`_.


Installation
------------

The best way to expose the JAR file is via HDFS as all your Hadoop nodes
will then have access to the new plugin.  Otherwise, you're free to
copy the MetlogHive jar file manually to all nodes to their local
filesystem.

Here's an example of dropping the JAR file into HDFS and making the
`exjson_tuple` function available for use ::

    $ hadoop dfs -mkdir /lib/hive
    $ hadoop dfs -put /opt/metlog/hadoop/hive/MetlogHive.jar /lib/hive/MetlogHive.jar
    $ hadoop dfs -ls /lib/hive/
    Found 1 items
    -rw-r--r--   1 vagrant supergroup     816600 2012-07-31 04:30 /lib/hive/MetlogHive.jar
    $ hive
    Hive history file=/tmp/vagrant/hive_job_log_vagrant_201207310430_1529091028.txt
    hive> add jar hdfs:///lib/hive/MetlogHive.jar;
    converting to local hdfs:///lib/hive/MetlogHive.jar
    Added /tmp/vagrant/hive_resources/MetlogHive.jar to class path
    Added resource: /tmp/vagrant/hive_resources/MetlogHive.jar
    hive> create temporary function exjson_tuple as 'org.mozilla.services.json.ExJSONTuple';
    OK
    Time taken: 1.839 seconds
    hive> 

If you have Hive scripts, you should put these two lines into the top of all your scripts.

    add jar hdfs:///lib/hive/MetlogHive.jar;
    create temporary function exjson_tuple as 'org.mozilla.services.json.ExJSONTuple';

This will expose a UDTF that will let you extract bits of a JSON blob and turn them into tuples for Hive.

Syntax
------

ExJSONTuple expects (json_blob, json_path1, json_path2, ... , json_pathN)

Each `json_path` is a dotted path notation string and can also use square bracket syntax to address items arrays.

Some examples should clear this up.

Here's a sample JSON blob that has been formatted so that it's easier to see what's going on ::

       {"severity":6,
        "timestamp":"2012-07-18T21:16:45.572517",
        "metlog_hostname":"aitc1.web.mtv1.dev.svc.mozilla.com",
        "fields":{"headers":{"path":"/1.0/11225/devices/",
                "host":"dev-aitc10.services.mozilla.com:443",
                "some_list": [5,10,15],
                "User-Agent":""},
        "threadlocal":{}},
        "metlog_pid":26482,
        "logger":"",
        "type":"wsgi",
        "payload":"",
        "env_version":"0.8"}


We can address the top level keys in this blob using paths like "severity", "timestamp" and "metlog_hostname". 
To address values in the `fields` section, you just have to use dot notation.

        ========================= =============================================
        Path                      Value
        ------------------------- ---------------------------------------------
        severity                       6  
        timestamp                      "2012-07-18T21:16:45.572517"
        metlog_hostname                "aitc1.web.mtv1.dev.svc.mozilla.com"
        fields.headers.path            "/1.0/11225/devices/"
        fields.host                    "dev-aitc10.services.mozilla.com:443"
        fields.some_list[2]            15
        ========================= =============================================


Example Usage
-------------

Sample output ::

    hive> select f1, f2, f3 from metrics m lateral view exjson_tuple(m.metlog_val, "fields.headers.host", "severity", "fields.headers.path") b as f1, f2, f3;
    Total MapReduce jobs = 1
    Launching Job 1 out of 1
    Number of reduce tasks is set to 0 since there's no reduce operator
    Starting Job = job_201207301923_0023, Tracking URL = http://localhost:50030/jobdetails.jsp?jobid=job_201207301923_0023
    Kill Command = /usr/lib/hadoop/bin/hadoop job  -Dmapred.job.tracker=localhost:8021 -kill job_201207301923_0023
    2012-07-31 03:41:14,114 Stage-1 map = 0%,  reduce = 0%
    2012-07-31 03:41:20,135 Stage-1 map = 100%,  reduce = 0%
    2012-07-31 03:41:24,154 Stage-1 map = 100%,  reduce = 100%
    Ended Job = job_201207301923_0023
    OK
    dev-aitc5.services.mozilla.com:443      6     /1.0/11224/apps/
    dev-aitc5.services.mozilla.com:443      6     /1.0/11224/apps/
    dev-aitc5.services.mozilla.com:443      6     /1.0/11224/devices/
    dev-aitc5.services.mozilla.com:443      6     /1.0/11224/devices/
    dev-aitc5.services.mozilla.com:443      6     /1.0/11224/apps/Mnw_2ofOKGhIpXSYLd0LfHSH-BY
    dev-aitc5.services.mozilla.com:443      6     /1.0/11224/apps/
    ...



Building the plugin
-------------------

Requirements:  

    * Java SDK.  (>= 1.6.0.33) 
    * Apache Ant (>= 1.8.2)
    * Apache Ivy (>= 2.3.0-rc1)
    * JUnit      (>= 4.10)

Assuming your Ivy and JUnit jar files are located in ~/.ant/lib, ant
should be able to run all targets in the build.xml file.  Your
~/.ant/lib should look something like this ::

    ~ > ls -l ~/.ant/lib
    total 2872
    -rw-r--r--  1 victorng  staff   1.2M 16 Apr 00:02 ivy-2.3.0-rc1.jar
    -rw-r--r--  1 victorng  staff   247K 30 Sep  2011 junit-4.10.jar
    ~ > 

Building a JAR file ::

    $ ant dist

Running tests ::

    $ ant test

Generating the Javadoc ::

    $ ant javadoc


CHANGES.rst
???
==================

- TODO



0.1 - 2012-07-30
================

- Initial release


src/org/mozilla/services/json/FastPath.java
package org.mozilla.services.json;

import java.util.StringTokenizer;

import org.apache.commons.lang.StringUtils;

import org.json.simple.parser.JSONParser;
import org.json.simple.JSONObject;
import org.json.simple.JSONAware;
import org.json.simple.JSONArray;
import org.json.simple.parser.ParseException;

public class FastPath {
    static JSONParser parser = new JSONParser();

    JSONAware coreObject;

    public FastPath(JSONAware aware) {
        this.coreObject = aware;
    }

    public FastPath(String json) {
        try {
            this.coreObject = (JSONAware) parser.parse(json);
        } catch (ParseException pe) {
            this.coreObject = null;
        }
    }

    /*
     * Fetch a path from the JSON object
     *
     * @param path A path using dot notation to index into a JSON object.
     *             Square brackets can be used to index into arrays.
     * @return The stringified value at the keypath.  NULL if the
     *         keypath walks is an invalid path
     */
    public String get(String path) {
        Object result;
        if (StringUtils.isBlank(path)) {
            result = this;
        }

        result = iterateThrough(this, path);
        if (result == null) {
            return null;
        }
        return result.toString();
    }

    private static Object iterateThrough (FastPath element, String path) {
        StringTokenizer st = new StringTokenizer(path, ".");

        Object traverser = element.coreObject;
        while (st.hasMoreTokens())
        { 
            String token = st.nextToken();
            if (isArray(token)) {
                traverser = getArrayElement((JSONObject)traverser, token);
            } else {
                traverser = ((JSONObject)traverser).get(token);
            }

            // Short circuit return a null if we fall off the JSON
            // object
            if (traverser == null) {
                return null;
            }
        }
        return traverser;
    }

    private static Object getArrayElement(JSONObject traverser,
            String token) {
        int index = Integer.valueOf(token.substring(token.indexOf("[") + 1,
                    token.indexOf("]")));
        String key = token.substring(0, token.indexOf("["));

        if ((index < 0) || (index >= ((JSONArray) traverser.get(key)).size()))
        {
            return null;
        }

        return ((JSONArray) traverser.get(key)).get(index);
    }

    private static boolean isArray(String token) {
        return token.contains("[")
            && token.contains("]")
            && (token.indexOf("[") < token.indexOf("]"))
            && StringUtils.isNumeric(token.substring(token.indexOf("[")+1,
                        token.indexOf("]")));
    }

    public String toString() {
        return this.coreObject.toString();
    }


}


src/org/mozilla/services/json/ExJSONTuple.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.mozilla.services.json;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.Map;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.apache.hadoop.hive.ql.exec.Description;
import org.apache.hadoop.hive.ql.exec.UDFArgumentException;
import org.apache.hadoop.hive.ql.metadata.HiveException;
import org.apache.hadoop.hive.ql.udf.generic.GenericUDTF;
import org.apache.hadoop.hive.serde.Constants;
import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspector;
import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspectorFactory;
import org.apache.hadoop.hive.serde2.objectinspector.StructObjectInspector;
import org.apache.hadoop.hive.serde2.objectinspector.primitive.PrimitiveObjectInspectorFactory;
import org.apache.hadoop.hive.serde2.objectinspector.primitive.StringObjectInspector;
import org.apache.hadoop.io.Text;
import org.json.simple.JSONObject;


/**
 * ExJSONTuple: this
 *
 */
@Description(name = "exjson_tuple",
    value = "_FUNC_(jsonStr, p1, p2, ..., pn) - like get_json_object, but it takes multiple names and return a tuple. " +
            "All the input parameters and output column types are string.")

public class ExJSONTuple extends GenericUDTF {

  private static Log LOG = LogFactory.getLog(ExJSONTuple.class.getName());

  int numCols;    // number of output columns
  String[] paths; // array of path expressions, each of which corresponds to a column
  Text[] retCols; // array of returned column values
  Text[] cols;    // object pool of non-null Text, avoid creating objects all the time
  Object[] nullCols; // array of null column values
  ObjectInspector[] inputOIs; // input ObjectInspectors
  boolean pathParsed = false;
  boolean seenErrors = false;

  //An LRU cache using a linked hash map
  static class HashCache<K, V> extends LinkedHashMap<K, V> {

    private static final int CACHE_SIZE = 16;
    private static final int INIT_SIZE = 32;
    private static final float LOAD_FACTOR = 0.6f;

    HashCache() {
      super(INIT_SIZE, LOAD_FACTOR);
    }

    private static final long serialVersionUID = 1;

    @Override
    protected boolean removeEldestEntry(Map.Entry<K, V> eldest) {
      return size() > CACHE_SIZE;
    }

  }

  static Map<String, FastPath> jsonObjectCache = new HashCache<String, FastPath>();

  @Override
  public void close() throws HiveException {
  }

  @Override
  public StructObjectInspector initialize(ObjectInspector[] args)
      throws UDFArgumentException {

    inputOIs = args;
    numCols = args.length - 1;

    if (numCols < 1) {
      throw new UDFArgumentException("exjson_tuple() takes at least two arguments: " +
              "the json string and a path expression");
    }

    for (int i = 0; i < args.length; ++i) {
      if (args[i].getCategory() != ObjectInspector.Category.PRIMITIVE ||
          !args[i].getTypeName().equals(Constants.STRING_TYPE_NAME)) {
        throw new UDFArgumentException("exjson_tuple()'s arguments have to be string type");
      }
    }

    seenErrors = false;
    pathParsed = false;
    paths = new String[numCols];
    cols = new Text[numCols];
    retCols = new Text[numCols];
    LOG.warn("setting up a new retCols array of size: " + numCols);
    nullCols = new Object[numCols];

    for (int i = 0; i < numCols; ++i) {
      cols[i] = new Text();
      retCols[i] = cols[i];
      nullCols[i] = null;
    }

    // construct output object inspector
    ArrayList<String> fieldNames = new ArrayList<String>(numCols);
    ArrayList<ObjectInspector> fieldOIs = new ArrayList<ObjectInspector>(numCols);
    for (int i = 0; i < numCols; ++i) {
      // column name can be anything since it will be named by UDTF as clause
      fieldNames.add("c" + i);
      // all returned type will be Text
      fieldOIs.add(PrimitiveObjectInspectorFactory.writableStringObjectInspector);
    }
    return ObjectInspectorFactory.getStandardStructObjectInspector(fieldNames, fieldOIs);
  }

  @Override
  public void process(Object[] o) throws HiveException {

    if (o[0] == null) {
      forward(nullCols);
      return;
    }
    // get the path expression for the 1st row only
    if (!pathParsed) {
      for (int i = 0;i < numCols; ++i) {
        paths[i] = ((StringObjectInspector) inputOIs[i+1]).getPrimitiveJavaObject(o[i+1]);
      }
      pathParsed = true;
    }

    String jsonStr = ((StringObjectInspector) inputOIs[0]).getPrimitiveJavaObject(o[0]);
    if (jsonStr == null) {
      forward(nullCols);
      return;
    }
    try {
      FastPath fastpath = jsonObjectCache.get(jsonStr);
      if (fastpath == null) {
        fastpath = new FastPath(jsonStr);
        jsonObjectCache.put(jsonStr, fastpath);
      }

      for (int i = 0; i < numCols; ++i) {
          if (fastpath.get(paths[i]) == null) {
              retCols[i] = null;
          } else {
              if (retCols[i] == null) {
                  retCols[i] = cols[i]; // use the object pool rather than creating a new object
              }
              retCols[i].set(fastpath.get(paths[i]));
              LOG.warn("adding string value ["+retCols[i]+"] @ index: " + i);
          }
      }
      forward(retCols);
      return;
    } catch (Throwable e) {
      LOG.error("JSON parsing/evaluation exception" + e);
      forward(nullCols);
    }
  }

  @Override
  public String toString() {
    return "exjson_tuple";
  }
}


src/org/mozilla/services/json/GenericUDTFJSONTuple.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.mozilla.services.json;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.Map;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.apache.hadoop.hive.ql.exec.Description;
import org.apache.hadoop.hive.ql.exec.UDFArgumentException;
import org.apache.hadoop.hive.ql.metadata.HiveException;
import org.apache.hadoop.hive.ql.udf.generic.GenericUDTF;
import org.apache.hadoop.hive.serde.Constants;
import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspector;
import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspectorFactory;
import org.apache.hadoop.hive.serde2.objectinspector.StructObjectInspector;
import org.apache.hadoop.hive.serde2.objectinspector.primitive.PrimitiveObjectInspectorFactory;
import org.apache.hadoop.hive.serde2.objectinspector.primitive.StringObjectInspector;
import org.apache.hadoop.io.Text;
import org.json.simple.JSONObject;


/**
 * GenericUDTFJSONTuple: this
 *
 */
@Description(name = "exjson_tuple",
    value = "_FUNC_(jsonStr, p1, p2, ..., pn) - like get_json_object, but it takes multiple names and return a tuple. " +
    		"All the input parameters and output column types are string.")

public class GenericUDTFJSONTuple extends GenericUDTF {

  private static Log LOG = LogFactory.getLog(GenericUDTFJSONTuple.class.getName());

  int numCols;    // number of output columns
  String[] paths; // array of path expressions, each of which corresponds to a column
  Text[] retCols; // array of returned column values
  Text[] cols;    // object pool of non-null Text, avoid creating objects all the time
  Object[] nullCols; // array of null column values
  ObjectInspector[] inputOIs; // input ObjectInspectors
  boolean pathParsed = false;
  boolean seenErrors = false;

  //An LRU cache using a linked hash map
  static class HashCache<K, V> extends LinkedHashMap<K, V> {

    private static final int CACHE_SIZE = 16;
    private static final int INIT_SIZE = 32;
    private static final float LOAD_FACTOR = 0.6f;

    HashCache() {
      super(INIT_SIZE, LOAD_FACTOR);
    }

    private static final long serialVersionUID = 1;

    @Override
    protected boolean removeEldestEntry(Map.Entry<K, V> eldest) {
      return size() > CACHE_SIZE;
    }

  }

  static Map<String, FastPath> jsonObjectCache = new HashCache<String, FastPath>();

  @Override
  public void close() throws HiveException {
  }

  @Override
  public StructObjectInspector initialize(ObjectInspector[] args)
      throws UDFArgumentException {

    inputOIs = args;
    numCols = args.length - 1;

    if (numCols < 1) {
      throw new UDFArgumentException("json_tuple() takes at least two arguments: " +
      		"the json string and a path expression");
    }

    for (int i = 0; i < args.length; ++i) {
      if (args[i].getCategory() != ObjectInspector.Category.PRIMITIVE ||
          !args[i].getTypeName().equals(Constants.STRING_TYPE_NAME)) {
        throw new UDFArgumentException("json_tuple()'s arguments have to be string type");
      }
    }

    seenErrors = false;
    pathParsed = false;
    paths = new String[numCols];
    cols = new Text[numCols];
    retCols = new Text[numCols];
    nullCols = new Object[numCols];

    for (int i = 0; i < numCols; ++i) {
      cols[i] = new Text();
      retCols[i] = cols[i];
      nullCols[i] = null;
    }

    // construct output object inspector
    ArrayList<String> fieldNames = new ArrayList<String>(numCols);
    ArrayList<ObjectInspector> fieldOIs = new ArrayList<ObjectInspector>(numCols);
    for (int i = 0; i < numCols; ++i) {
      // column name can be anything since it will be named by UDTF as clause
      fieldNames.add("c" + i);
      // all returned type will be Text
      fieldOIs.add(PrimitiveObjectInspectorFactory.writableStringObjectInspector);
    }
    return ObjectInspectorFactory.getStandardStructObjectInspector(fieldNames, fieldOIs);
  }

  @Override
  public void process(Object[] o) throws HiveException {

    if (o[0] == null) {
      forward(nullCols);
      return;
    }
    // get the path expression for the 1st row only
    if (!pathParsed) {
      for (int i = 0;i < numCols; ++i) {
        paths[i] = ((StringObjectInspector) inputOIs[i+1]).getPrimitiveJavaObject(o[i+1]);
      }
      pathParsed = true;
    }

    String jsonStr = ((StringObjectInspector) inputOIs[0]).getPrimitiveJavaObject(o[0]);
    if (jsonStr == null) {
      forward(nullCols);
      return;
    }
    try {
      FastPath fastpath = jsonObjectCache.get(jsonStr);
      if (fastpath == null) {
        fastpath = new FastPath(jsonStr);
        jsonObjectCache.put(jsonStr, fastpath);
      }

      for (int i = 0; i < numCols; ++i) {
          // TODO: this can be optimized a bit so that we don't double
          // fetch
        if (fastpath.get(paths[i]) != null) {
          retCols[i] = null;
        } else {
          if (retCols[i] == null) {
            retCols[i] = cols[i]; // use the object pool rather than creating a new object
          }
          retCols[i].set(fastpath.get(paths[i]));
        }
      }
      forward(retCols);
      return;
    } catch (Throwable e) {
      LOG.error("JSON parsing/evaluation exception" + e);
      forward(nullCols);
    }
  }

  @Override
  public String toString() {
    return "json_tuple";
  }
}
